File: server/src/pulse.js
'use strict';

const { db } = require('./db');

/**
 * Market Pulse — the institutional-style landing screen. Everything here is
 * derived fresh from `listings` on every request (no separate snapshot
 * table to keep in sync) so it's always consistent with the Overview/
 * Vendor/Volume pages; "stored over time" comes from the listings table
 * itself accumulating across imports, not from a bespoke pulse-history log.
 *
 * "Now" is the latest record_date in the dataset, not wall-clock time —
 * same convention as pricing.js, since this data's own dates run into
 * 2026+ regardless of when the app is actually opened.
 */

const PERIOD_DAYS = 60;

function daysBetween(aIso, bIso) {
  return (new Date(bIso).getTime() - new Date(aIso).getTime()) / 86400000;
}

function addDays(iso, days) {
  const d = new Date(iso);
  d.setUTCDate(d.getUTCDate() + days);
  return d.toISOString().slice(0, 10);
}

function getLatestDate() {
  // Exclude date_flagged rows (e.g. the known "2206" instead of "2026" typo,
  // see CLAUDE.md §1) — a bad future date would otherwise poison every
  // "now"-relative calculation (recency weighting, days-on-market, periods).
  const row = db.prepare("SELECT MAX(record_date) as d FROM listings WHERE date_flagged = 0").get();
  return row?.d || new Date().toISOString().slice(0, 10);
}

function usableRows() {
  return db.prepare("SELECT * FROM listings WHERE price_status NOT IN ('missing')").all();
}

function inRange(row, start, end) {
  return row.record_date && row.record_date >= start && row.record_date <= end;
}

function sum(arr, fn) {
  return arr.reduce((s, x) => s + (fn(x) ?? 0), 0);
}

function pctChange(curr, prior) {
  if (prior === null || prior === undefined || prior === 0) return null;
  if (curr === null || curr === undefined) return null;
  return ((curr - prior) / Math.abs(prior)) * 100;
}

function periodKpis(rows, market) {
  const txns = rows.filter((r) => r.record_type === 'transaction' && (!market || r.market === market));
  const avails = rows.filter((r) => r.record_type === 'availability' && (!market || r.market === market));
  const saleTxns = txns.filter((r) => r.price_deal_type === 'sale' && r.price_unit === 'total' && r.price_amount != null);
  const totalVolume = sum(saleTxns, (r) => r.price_amount);
  return {
    dealsClosed: txns.length,
    newAvailabilities: avails.length,
    totalVolume,
    avgDealSize: saleTxns.length ? totalVolume / saleTxns.length : null,
    saleCompCount: saleTxns.length,
  };
}

function avgPerAcre(rows) {
  const withAcre = rows.filter((r) => r.price_deal_type === 'sale' && r.price_unit === 'total' && r.acres && r.price_amount != null);
  if (withAcre.length === 0) return null;
  return sum(withAcre, (r) => r.price_amount / r.acres) / withAcre.length;
}

function buildMonthlySeries(rows, monthsBack = 9) {
  const now = getLatestDate();
  const months = [];
  for (let i = monthsBack - 1; i >= 0; i--) {
    const d = new Date(now);
    d.setUTCMonth(d.getUTCMonth() - i);
    months.push(d.toISOString().slice(0, 7));
  }
  const byMonth = new Map(months.map((m) => [m, { month: m, transactions: 0, availabilities: 0, volume: 0 }]));
  for (const r of rows) {
    if (!r.record_date) continue;
    const m = r.record_date.slice(0, 7);
    if (!byMonth.has(m)) continue;
    const bucket = byMonth.get(m);
    if (r.record_type === 'transaction') {
      bucket.transactions += 1;
      if (r.price_deal_type === 'sale' && r.price_unit === 'total' && r.price_amount) bucket.volume += r.price_amount;
    } else {
      bucket.availabilities += 1;
    }
  }
  return months.map((m) => byMonth.get(m));
}

function buildInsights(rows, current, prior, latestBatch) {
  const insights = [];

  // Biggest deal this period
  const currentSales = current.filter((r) => r.record_type === 'transaction' && r.price_deal_type === 'sale' && r.price_unit === 'total' && r.price_amount);
  if (currentSales.length) {
    const biggest = currentSales.reduce((a, b) => (b.price_amount > a.price_amount ? b : a));
    insights.push({
      type: 'biggest_deal',
      severity: 'info',
      text: `Largest sale in the last ${PERIOD_DAYS} days: ${biggest.address} (${biggest.municipality}) at $${Math.round(biggest.price_amount).toLocaleString()}.`,
    });
  }

  // Price movers per market+type (avg $/acre, current vs prior)
  const MARKETS = [...new Set(rows.map((r) => r.market))];
  const TYPES = ['Low Coverage', 'Land'];
  let biggestMove = null;
  for (const m of MARKETS) {
    for (const t of TYPES) {
      const curr = avgPerAcre(current.filter((r) => r.market === m && r.type_of_land_normalized === t));
      const pri = avgPerAcre(prior.filter((r) => r.market === m && r.type_of_land_normalized === t));
      const change = pctChange(curr, pri);
      if (change !== null && (!biggestMove || Math.abs(change) > Math.abs(biggestMove.change))) {
        biggestMove = { market: m, type: t, change, curr, pri };
      }
    }
  }
  if (biggestMove) {
    const dir = biggestMove.change >= 0 ? 'up' : 'down';
    insights.push({
      type: 'price_move',
      severity: Math.abs(biggestMove.change) > 15 ? 'high' : 'info',
      text: `${biggestMove.market} ${biggestMove.type} avg $/acre is ${dir} ${Math.abs(biggestMove.change).toFixed(0)}% vs the prior ${PERIOD_DAYS} days ($${Math.round(biggestMove.pri).toLocaleString()} → $${Math.round(biggestMove.curr).toLocaleString()}).`,
    });
  }

  // Repeat vendors (overall, not period-scoped — pattern signal)
  const vendorCounts = new Map();
  for (const r of rows) {
    if (r.record_type !== 'transaction' || !r.vendor_normalized || r.vendor_normalized === 'Unknown') continue;
    vendorCounts.set(r.vendor_normalized, (vendorCounts.get(r.vendor_normalized) || 0) + 1);
  }
  const repeatVendors = [...vendorCounts.entries()].filter(([, n]) => n >= 2).sort((a, b) => b[1] - a[1]);
  if (repeatVendors.length) {
    insights.push({
      type: 'repeat_vendor',
      severity: 'info',
      text: `${repeatVendors[0][0]} has ${repeatVendors[0][1]} deals on file — worth a closer look as a repeat principal.${repeatVendors.length > 1 ? ` (${repeatVendors.length} vendors total with 2+ deals.)` : ''}`,
    });
  }

  // Longest-active availability still on market
  const activeAvail = rows.filter((r) => r.record_type === 'availability' && r.first_seen_date);
  if (activeAvail.length) {
    const now = getLatestDate();
    const withAge = activeAvail.map((r) => ({ r, age: daysBetween(r.first_seen_date, now) }));
    const oldest = withAge.reduce((a, b) => (b.age > a.age ? b : a));
    if (oldest.age >= 14) {
      insights.push({
        type: 'stale_listing',
        severity: oldest.age > 90 ? 'high' : 'info',
        text: `${oldest.r.address} (${oldest.r.municipality}) has been tracked as available for ${Math.round(oldest.age)} days — longest currently on file.`,
      });
    }
  }

  // What changed since the last import
  if (latestBatch) {
    const newInLastBatch = rows.filter((r) => r.first_seen_batch_id === latestBatch.id);
    const updatedInLastBatch = rows.filter((r) => r.last_seen_batch_id === latestBatch.id && r.first_seen_batch_id !== latestBatch.id);
    if (newInLastBatch.length || updatedInLastBatch.length) {
      insights.push({
        type: 'last_import',
        severity: 'info',
        text: `Last import ("${latestBatch.filename}") added ${newInLastBatch.length} new listing(s) and refreshed ${updatedInLastBatch.length} already-tracked one(s).`,
      });
    }
  }

  // Data quality
  const needsReview = rows.filter((r) => r.needs_review).length;
  if (needsReview > 0) {
    insights.push({
      type: 'data_quality',
      severity: needsReview > 20 ? 'high' : 'medium',
      text: `${needsReview} listing(s) still need review before they're fully trusted in the averages above.`,
    });
  }

  return insights;
}

function getPulse() {
  const rows = usableRows();
  const now = getLatestDate();
  const periodStart = addDays(now, -PERIOD_DAYS);
  const priorStart = addDays(now, -PERIOD_DAYS * 2);

  const current = rows.filter((r) => inRange(r, periodStart, now));
  const prior = rows.filter((r) => inRange(r, priorStart, periodStart));

  const overall = { current: periodKpis(current), prior: periodKpis(prior) };
  const byMarket = {};
  for (const m of [...new Set(rows.map((r) => r.market))]) {
    byMarket[m] = { current: periodKpis(current, m), prior: periodKpis(prior, m) };
  }

  const latestBatch = db.prepare('SELECT * FROM import_batches ORDER BY id DESC LIMIT 1').get();
  // date_flagged rows (e.g. the "2206" typo) are excluded here even though
  // they're still counted everywhere else — a bad date would otherwise sort
  // to the top of "recent" activity and read as misleadingly current.
  const recentActivity = db
    .prepare("SELECT * FROM listings WHERE date_flagged = 0 ORDER BY record_date DESC, updated_at DESC LIMIT 15")
    .all();

  return {
    asOf: now,
    periodDays: PERIOD_DAYS,
    overall,
    byMarket,
    trend: buildMonthlySeries(rows),
    insights: buildInsights(rows, current, prior, latestBatch),
    recentActivity,
    totalListings: rows.length,
    totalTransactions: rows.filter((r) => r.record_type === 'transaction').length,
    totalAvailabilities: rows.filter((r) => r.record_type === 'availability').length,
    needsReviewCount: rows.filter((r) => r.needs_review).length,
    lastImport: latestBatch || null,
  };
}

module.exports = { getPulse };

File: streamlit_app/lib/pulse.py
"""Market Pulse — ported from pulse.js. Everything derived fresh from
`listings` on every call. "Now" is the latest *unflagged* record_date in
the dataset (excluding the known '2206' typo pattern — see get_latest_date).
"""

import datetime as dt

from . import db

PERIOD_DAYS = 60


def get_latest_date():
    row = db.query_one("SELECT MAX(record_date) as d FROM listings WHERE date_flagged = 0")
    return row["d"] if row and row["d"] else dt.datetime.now().strftime("%Y-%m-%d")


def _add_days(iso, days):
    d = dt.datetime.strptime(iso, "%Y-%m-%d") + dt.timedelta(days=days)
    return d.strftime("%Y-%m-%d")


def _days_between(a_iso, b_iso):
    a = dt.datetime.strptime(a_iso, "%Y-%m-%d")
    b = dt.datetime.strptime(b_iso, "%Y-%m-%d")
    return (b - a).days


def _usable_rows():
    return [dict(r) for r in db.query("SELECT * FROM listings WHERE price_status != 'missing'")]


def _in_range(row, start, end):
    return row["record_date"] and start <= row["record_date"] <= end


def _pct_change(curr, prior):
    if prior in (None, 0) or curr is None:
        return None
    return ((curr - prior) / abs(prior)) * 100


def _period_kpis(rows, market=None):
    txns = [r for r in rows if r["record_type"] == "transaction" and (not market or r["market"] == market)]
    avails = [r for r in rows if r["record_type"] == "availability" and (not market or r["market"] == market)]
    sale_txns = [r for r in txns if r["price_deal_type"] == "sale" and r["price_unit"] == "total" and r["price_amount"] is not None]
    total_volume = sum(r["price_amount"] for r in sale_txns)
    return {
        "deals_closed": len(txns),
        "new_availabilities": len(avails),
        "total_volume": total_volume,
        "avg_deal_size": total_volume / len(sale_txns) if sale_txns else None,
        "sale_comp_count": len(sale_txns),
    }


def _avg_per_acre(rows):
    with_acre = [r for r in rows if r["price_deal_type"] == "sale" and r["price_unit"] == "total" and r["acres"] and r["price_amount"] is not None]
    if not with_acre:
        return None
    return sum(r["price_amount"] / r["acres"] for r in with_acre) / len(with_acre)


def _build_monthly_series(rows, months_back=9):
    now = get_latest_date()
    y, m = int(now[:4]), int(now[5:7])
    months = []
    for i in range(months_back - 1, -1, -1):
        mm = m - i
        yy = y
        while mm <= 0:
            mm += 12
            yy -= 1
        months.append(f"{yy:04d}-{mm:02d}")

    by_month = {mo: {"month": mo, "transactions": 0, "availabilities": 0, "volume": 0} for mo in months}
    for r in rows:
        if not r["record_date"]:
            continue
        mo = r["record_date"][:7]
        if mo not in by_month:
            continue
        b = by_month[mo]
        if r["record_type"] == "transaction":
            b["transactions"] += 1
            if r["price_deal_type"] == "sale" and r["price_unit"] == "total" and r["price_amount"]:
                b["volume"] += r["price_amount"]
        else:
            b["availabilities"] += 1
    return [by_month[mo] for mo in months]


def _build_insights(rows, current, prior, latest_batch):
    insights = []

    current_sales = [r for r in current if r["record_type"] == "transaction" and r["price_deal_type"] == "sale" and r["price_unit"] == "total" and r["price_amount"]]
    if current_sales:
        biggest = max(current_sales, key=lambda r: r["price_amount"])
        insights.append({
            "type": "biggest_deal", "severity": "info",
            "text": f"Largest sale in the last {PERIOD_DAYS} days: {biggest['address']} ({biggest['municipality']}) at ${round(biggest['price_amount']):,}.",
        })

    markets = sorted({r["market"] for r in rows})
    types = ["Low Coverage", "Land"]
    biggest_move = None
    for mkt in markets:
        for typ in types:
            curr = _avg_per_acre([r for r in current if r["market"] == mkt and r["type_of_land_normalized"] == typ])
            pri = _avg_per_acre([r for r in prior if r["market"] == mkt and r["type_of_land_normalized"] == typ])
            change = _pct_change(curr, pri)
            if change is not None and (biggest_move is None or abs(change) > abs(biggest_move["change"])):
                biggest_move = {"market": mkt, "type": typ, "change": change, "curr": curr, "pri": pri}
    if biggest_move:
        direction = "up" if biggest_move["change"] >= 0 else "down"
        insights.append({
            "type": "price_move", "severity": "high" if abs(biggest_move["change"]) > 15 else "info",
            "text": f"{biggest_move['market']} {biggest_move['type']} avg $/acre is {direction} {abs(biggest_move['change']):.0f}% "
                    f"vs the prior {PERIOD_DAYS} days (${round(biggest_move['pri']):,} → ${round(biggest_move['curr']):,}).",
        })

    vendor_counts = {}
    for r in rows:
        if r["record_type"] != "transaction" or not r["vendor_normalized"] or r["vendor_normalized"] == "Unknown":
            continue
        vendor_counts[r["vendor_normalized"]] = vendor_counts.get(r["vendor_normalized"], 0) + 1
    repeat_vendors = sorted([(k, v) for k, v in vendor_counts.items() if v >= 2], key=lambda kv: -kv[1])
    if repeat_vendors:
        extra = f" ({len(repeat_vendors)} vendors total with 2+ deals.)" if len(repeat_vendors) > 1 else ""
        insights.append({
            "type": "repeat_vendor", "severity": "info",
            "text": f"{repeat_vendors[0][0]} has {repeat_vendors[0][1]} deals on file — worth a closer look as a repeat principal.{extra}",
        })

    active_avail = [r for r in rows if r["record_type"] == "availability" and r["first_seen_date"]]
    if active_avail:
        now = get_latest_date()
        with_age = [(r, _days_between(r["first_seen_date"], now)) for r in active_avail]
        oldest, age = max(with_age, key=lambda t: t[1])
        if age >= 14:
            insights.append({
                "type": "stale_listing", "severity": "high" if age > 90 else "info",
                "text": f"{oldest['address']} ({oldest['municipality']}) has been tracked as available for {age} days — longest currently on file.",
            })

    if latest_batch:
        new_in_last = [r for r in rows if r["first_seen_batch_id"] == latest_batch["id"]]
        updated_in_last = [r for r in rows if r["last_seen_batch_id"] == latest_batch["id"] and r["first_seen_batch_id"] != latest_batch["id"]]
        if new_in_last or updated_in_last:
            insights.append({
                "type": "last_import", "severity": "info",
                "text": f'Last import ("{latest_batch["filename"]}") added {len(new_in_last)} new listing(s) and refreshed {len(updated_in_last)} already-tracked one(s).',
            })

    needs_review = sum(1 for r in rows if r["needs_review"])
    if needs_review > 0:
        insights.append({
            "type": "data_quality", "severity": "high" if needs_review > 20 else "medium",
            "text": f"{needs_review} listing(s) still need review before they're fully trusted in the averages above.",
        })

    return insights


def get_pulse():
    rows = _usable_rows()
    now = get_latest_date()
    period_start = _add_days(now, -PERIOD_DAYS)
    prior_start = _add_days(now, -PERIOD_DAYS * 2)

    current = [r for r in rows if _in_range(r, period_start, now)]
    prior = [r for r in rows if _in_range(r, prior_start, period_start)]

    overall = {"current": _period_kpis(current), "prior": _period_kpis(prior)}
    by_market = {}
    for mkt in sorted({r["market"] for r in rows}):
        by_market[mkt] = {"current": _period_kpis(current, mkt), "prior": _period_kpis(prior, mkt)}

    latest_batch_row = db.query_one("SELECT * FROM import_batches ORDER BY id DESC LIMIT 1")
    latest_batch = dict(latest_batch_row) if latest_batch_row else None

    recent_rows = db.query(
        "SELECT * FROM listings WHERE date_flagged = 0 ORDER BY record_date DESC, updated_at DESC LIMIT 15"
    )
    recent_activity = [dict(r) for r in recent_rows]

    return {
        "as_of": now,
        "period_days": PERIOD_DAYS,
        "overall": overall,
        "by_market": by_market,
        "trend": _build_monthly_series(rows),
        "insights": _build_insights(rows, current, prior, latest_batch),
        "recent_activity": recent_activity,
        "total_listings": len(rows),
        "total_transactions": sum(1 for r in rows if r["record_type"] == "transaction"),
        "total_availabilities": sum(1 for r in rows if r["record_type"] == "availability"),
        "needs_review_count": sum(1 for r in rows if r["needs_review"]),
        "last_import": latest_batch,
    }
